preciceconfigchecker/rule.py

```diff
@@ -1,37 +1,121 @@
-        else:
-            self.followed = result
-
-    def get_result(self) -> str:
-        if (self.followed):
-            return ""
-        Rule.numbers_not_followed += 1
-        out:str = f"[{Rule.numbers_not_followed:2}.]" + "[" + self.severity.value + "][" + self.name + "] : " + self.message
-        if (len(self.possible_solutions) > 0):
-            out += "\n\t[Possible Solution] : " + self.possible_solutions
-        return out
\ No newline at end of file
+        self.problem = problem
+        self.rule:str
+        self.formatted_violations:List[str] = []
+        self.violations:List[tuple] = []
+
+    @abstractmethod
+    def check_method(self) -> List[tuple]:
+        """
+        @abstractmethod: Defines how a 'Rule' should be checked
+
+        Returns:
+            List[tuple]: of all violations where the 'Rule' was not satisfied.
+
+        Tip: When overwriting, define the type of the tuple for easier use.
+        Tip: If the tuple only has one argument, put a comma ( , ) after it so that it is recognized as a tuple. Python specific!
+        """
+        pass
+
+    @abstractmethod
+    def format_explanation(self, data:tuple) -> str:
+        """
+        @abstractmethod: Formats the explanation of a violation
+
+        Args:
+            data (tuple): to use for formatting
+
+        Returns:
+            str: formatted
+
+        Tip: When overwriting, multiple arguments can be passed directly, rather than a tuple of arguments.
+        Tip: Define the abstract 'check_method' beforehand. Define the types of the tuple there and use these types for the arguments here.
+        """
+        pass
+    
+    @abstractmethod
+    def format_possible_solutions(self, data:tuple) -> List[str]:
+        """
+        @abstractmethod: Formats multiple possible solutions to a violation
+
+        Args:
+            data (tuple): to use for formatting
+
+        Returns:
+            List[str]: of formatted possible solutions
+
+        Tip: When overwriting, multiple arguments can be passed directly, rather than a tuple of arguments.
+        Tip: Define the abstract 'check_method' beforehand. Define the types of the tuple there and use these types for the arguments here.
+        """
+        pass
+    
+    def satisfied(self) -> bool:
+        """
+        Shows when the 'Rule' is satisfied
+
+        Returns:
+            bool: TRUE if there are no violations after the check
+        """
+        return (len(self.violations) == 0)
+    
+    def can_print(self) -> bool:
+        """
+        Shows if the 'Rule' can print formatted violations
+
+        Returns:
+            bool: TRUE if formatted violations are available
+        """
+        return (len(self.formatted_violations) > 0)
+    
+    def check(self) -> None:
+        """
+        Runs the abstract 'check_method' and saves the violations
+        """
+        self.violations = self.check_method()
```

Comment:
That's what I did first.

Then the following idea came:
In the end you should have better control of the data types when overwriting. If you just push something onto a list, you either don't recognize the data types quickly enough or not at all. Therefore, this is intended to encourage the user to define data types and to use exactly these data types in the formatting methods. This will also be easier for other people to use in the future.

Reply:
Done